### PR TITLE
ASoC: SOF: Fix HDMI playback S3 failure

### LINKED ISSUE #489: PM: Link DMA fails to resume after system suspend
This issue only happens with aplay, no such issue if use browser player to play audio via HDMI or I2S.
Plus, we only observe this issue on HDMI with 0.55 recipe.
But with PV recipe, both hdmi and i2s can reproduce this issue.

How to reproduce:
1. open one terminal and play audio via hdmi monitor.
    aplay -Dhw:0,5 -f dat -c 2 audio.wav
2. open another terminal and run suspend/resume test
    suspend_stress_test 
3. then system will reboot

No log is available.

PV recipe:
kernel: cros-sof-v4.14-rebase
sof/tplg: glk-005-drop-stable

### PR DESCRIPTION
**Pull request by libinyang**
After S3 from HDMI playback, BE's cpu_dai hw_params() should be
called properly to set the HDA link registers. Otherwise, the
link DMA transfer will be broken.

it fixes #489

Signed-off-by: Libin Yang <libin.yang@intel.com>

### CODE REVIEW

**Comment by ranj063** [comment]
@libinyang a few questions for my understanding:
1. hw_params callback here calls hda_link_hw_params()?
2. If so, the difference between what I had in my patch and this PR is the fact that hw_params for the BE cpu dai is getting set during the FE hw_params callback whereas I had set hw_params in the RESUME trigger callback.
3. If this is only needed for HDMI, is this the right place for it?

**Comment by libinyang** [comment]
@ranj063 
> 
> 1. hw_params callback here calls hda_link_hw_params()?

Yes.

> 2. If so, the difference between what I had in my patch and this PR is the fact that hw_params for the BE cpu dai is getting set during the FE hw_params callback whereas I had set hw_params in the RESUME trigger callback.

Yes, you are right. However there are 2 issues in your patch:
1) the call sequence. In your patch, it will be too late to call hda_link_hw_params(). DMA already started when hda_link_hw_params() is called. It took me a lot time to figure out this issue. Actually, there is no requirement for the register setting sequence in the SPEC. But based on my test, it does have such requirement. And I think such requirement is easy to understand.

2) the hw_params is wrong. In your patch, you used the FE raw params, however we should use the fixed up params.

> 3. If this is only needed for HDMI, is this the right place for it?

Good question. I think that setting BE hw_params is a general requirement, not only for HDMI. What do you think?

**Comment by ranj063** [comment]
@libinyang in 1. in your comment, do you mean to say the BE hw_params should be set before host DMA has been triggered?

As for the last question, you mean to say that the BE hw_params is not needed for SSP because these dont set their ops?

**Comment by libinyang** [comment]
@ranj063 
> @libinyang in 1. in your comment, do you mean to say the BE hw_params should be set before host DMA has been triggered?

Yes.

> 
> As for the last question, you mean to say that the BE hw_params is not needed for SSP because these dont set their ops?

My understanding is that: hw_params is used to setup the hardware, for example, the registers. So if the driver needs hw_params() when playback, we will still need to call hw_params in playback resume.

**Comment by plbossart** [comment]
@libinyang I don't see in what way the suggested patch is related to HDMI, this is a generic link DMA management issue that should also affect the HDaudio codec use. I've seen too many patches related to this link DMA, we need @RanderWang to chime in and have an agreed design.

**Comment by libinyang** [comment]
@plbossart  Yes, it is related to link DMA. And all HDaudio codec used by sof need such patch as they need link DMA to be set correctly.

> @libinyang I don't see in what way the suggested patch is related to HDMI, this is a generic link DMA management issue that should also affect the HDaudio codec use. I've seen too many patches related to this link DMA, we need @RanderWang to chime in and have an agreed design.

**Comment by lgirdwood** [comment]
@tlauda fyi, should fix https://github.com/thesofproject/sof/issues/1152

**Comment by libinyang** [comment]
@mmaka1 @tlauda 
Could you please help review this patch? Thanks.
After resume, the original code doesn't set the PP and ML register for the link DMA, including setting the stream tag in pplcctrl, format in pplcfmt. And the corresponding setting in LOSDIVx in ML registers. Could you please confirm if it is necessary? Thanks.

**Comment by libinyang** [comment]
@lgirdwood @tlauda 

This patch is not fixing the latest FW issue, but fixing the bug in driver for HDA codec audio in playback:
After resume, the original codec doesn't set the link DMA correctly.

> @tlauda fyi, should fix [thesofproject/sof#1152](https://github.com/thesofproject/sof/issues/1152)

**Comment by keqiaozhang** [comment]
I tested this PR on WHL, it can reduce the probability of this issue(100% ->~10%). But raise a new issue is that HDMI playback will be muted after resuming.

**Comment by libinyang** [comment]
@RanderWang 

Suppose it should be OK. As the params I used is the fixed up one.
I will test it. Ranjani suggested using prepare() for it. I think it's a good suggestion.

> Did you test it on HDA codec with all pcm format ? S16, 24, 32 ?
> According to dpcm_be_dai_hw_params, there are many corner cases.
> Now the flow is copied to here without these checks.
> As I know, hw_params would be fixup for sof for S16 and S32.
> 
> ```
> for_each_dpcm_be(fe, stream, dpcm) {
> 		if (!snd_soc_dpcm_be_can_update(fe, be, stream))
>                         ...
>                 if (be->dai_link->be_hw_params_fixup)
>                          ...
>                 if (!snd_soc_dpcm_can_be_params(fe, be, stream))
>                           ...
> 
>                  soc_pcm_hw_params(be_substream, &dpcm->hw_params);
> }
> ```
> To make it simple, just restore it at be link. It is complex to do it at FE.

**Comment by libinyang** [comment]
> To make it simple, just restore it at be link. It is complex to do it at FE.

@RanderWang We did do it in BE. But as in my comments before, it doesn't work.

**Comment by wenqingfu** [comment]
> @lgirdwood @tlauda
> 
> This patch is not fixing the latest FW issue, but fixing the bug in driver for HDA codec audio in playback:

Though thesofproject/sof#1152 is reported in firmware repository, it's not necessary a FW issue. Do you see they are not the same issue?

**Comment by libinyang** [comment]
> > Did you test it on HDA codec with all pcm format ? S16, 24, 32 ?

S16, 24, 32 are all tested. All works well.

**Comment by libinyang** [comment]
There are another bug in HDMI playback S3. It's a totally different bug from the bug this patch try to fix.

Please see https://github.com/thesofproject/linux/issues/737 for details.

**Comment by libinyang** [comment]
@wenqingfu You are right. They should be the same issue. Sometimes, from the first sight, it's a FW bug as FW print out the error messages. But as driver and FW are working so tightly, sometimes the FW unstability is caused by driver and vice versa. 

> Though [thesofproject/sof#1152](https://github.com/thesofproject/sof/issues/1152) is reported in firmware repository, it's not necessary a FW issue. Do you see they are not the same issue?

**Comment by RanderWang** [comment]
> To make it simple, just restore it at be link. It is complex to do it at FE.
> 
> @RanderWang We did do it in BE. But as in my comments before, it doesn't work.

I will discuss with you about your solution in BE tommorow. I made it work successfully one month ago .

**Comment by plbossart** [comment]
Guys, I will officially ignore those PRs until
a) there is an agreement between developers on the design
b) we have a fix that results in valid audio being rendered. If this 'fix' only results in muted audio it's not a fix. I don't want to track isolated but related PRs. Give me *one* PR that fixes HDMI problems.

**Comment by libinyang** [comment]
A patch to fix the no sound after HDMI playback S3 is submitted to the community.
The patch can be found at:
https://github.com/thesofproject/linux/pull/744

**Comment by RanderWang** [comment]
I discussed with Libin, checked the log and found that:
now the calling sequence is:
````
              FE resumed
              pcm start
              BE resumed
````
It is very strange, so Libin did it in FE resuming function.
the expected behavior should be
````
              FE resumed
              BE resumed
              pcm start
````

**Comment by libinyang** [comment]
> I discussed with Libin, checked the log and found that:
> now the calling sequence is:
> 
> ```
>               FE resumed
>               pcm start
>               BE resumed
> ```
> It is very strange, so Libin did it in FE resuming function.
> the expected behavior should be
> 
> ```
>               FE resumed
>               BE resumed
>               pcm start
> ```

Yes, this is why I did in FE. Thanks for review, Rander.

**Comment by libinyang** [comment]
@ranj063 @RanderWang 
Do we have agreement this patch is OK?

**Comment by plbossart** [comment]
@libinyang I don't understand neither your fix nor how it changes the calling sequence. It's very very hard to review those patches, please try to make the explanations as simple as possible so that folks can actually comment. The more obscure/specialized the explanations are the more time it'll consume for people to agree on the suggested direction.

File: sound/soc/sof/pcm.c [hunk 1]
```diff
@@ -216,8 +216,29 @@ static int sof_restore_hw_params(struct snd_pcm_substream *substream,
 				 struct snd_sof_dev *sdev)
 {
 	snd_pcm_uframes_t host = 0;
+	struct snd_soc_dpcm *dpcm;
+	struct snd_soc_pcm_runtime *fe = substream->private_data;
 	u64 host_posn;
 	int ret = 0;
+	int stream = substream->stream;
+
+	for_each_dpcm_be(fe, stream, dpcm) {
```
Comment (lgirdwood) [inline sound/soc/sof/pcm.c]: Please comment what we are doing here e.g. 
```
/* restore hw_params for each BE if needed */

File: sound/soc/sof/pcm.c [hunk 2]
```diff
@@ -216,8 +216,29 @@ static int sof_restore_hw_params(struct snd_pcm_substream *substream,
...
+		struct snd_soc_pcm_runtime *be = dpcm->be;
+		struct snd_pcm_substream *be_substream =
+			snd_soc_dpcm_get_substream(be, stream);
+		struct snd_soc_dai *dai = be->cpu_dai;
+
+		if (dai->driver->ops->hw_params) {
+			ret = dai->driver->ops->hw_params(be_substream,
```
Comment (lgirdwood) [inline sound/soc/sof/pcm.c]: Please check there is not a wrapper function for this indirection.

**Comment by ranj063** [comment]
@lgirdwood @plbossart @libinyang 

One thing I noticed was that all the FE dai link definitions in the machine driver set the trigger order to 
SND_SOC_DPCM_TRIGGER_POST. 

But SOF uses the default order which is SND_SOC_DPCM_TRIGGER_PRE. Could this be the reason that we're facing the issue? If so, can we use SND_SOC_DPCM_TRIGGER_POST and then restore hw_params in the BE instead?

File: sound/soc/sof/pcm.c (hunk 1 above)
Comment (libinyang) [reply sound/soc/sof/pcm.c]: I will do it. Thanks.

File: sound/soc/sof/pcm.c (hunk 2 above)
Comment (libinyang) [reply sound/soc/sof/pcm.c]: There is a function doing the similar things in asoc core. But it is not exported. Or let's export it?

**Comment by libinyang** [comment]
@ranj063 This is another solution. But our sof topology doesn't support it. We need a lot of thing to do. And what's more, this will change the default behavior which is different from before, we need do full test for the new behavior.

> @lgirdwood @plbossart @libinyang
> 
> One thing I noticed was that all the FE dai link definitions in the machine driver set the trigger order to
> SND_SOC_DPCM_TRIGGER_POST.
> 
> But SOF uses the default order which is SND_SOC_DPCM_TRIGGER_PRE. Could this be the reason that we're facing the issue? If so, can we use SND_SOC_DPCM_TRIGGER_POST and then restore hw_params in the BE instead?

**Comment by libinyang** [comment]
@ranj063 
after a second thought. You suggestion may be reasonable. We can set it in BE, and pass the trigger order to FE when BE and FE connect. I will have a try.

> @lgirdwood @plbossart @libinyang
> 
> One thing I noticed was that all the FE dai link definitions in the machine driver set the trigger order to
> SND_SOC_DPCM_TRIGGER_POST.
> 
> But SOF uses the default order which is SND_SOC_DPCM_TRIGGER_PRE. Could this be the reason that we're facing the issue? If so, can we use SND_SOC_DPCM_TRIGGER_POST and then restore hw_params in the BE instead?

File: sound/soc/sof/pcm.c (hunk 2 above)
Comment (lgirdwood) [reply sound/soc/sof/pcm.c]: Easier to send a patch that exports rather than re-implementing.

**Comment by lgirdwood** [comment]
@ranj063 trigger ordering was added (by me) to help on system like OMAP4 where there was complex ordering of HW programming that had to be shared between host and DSP. @libinyang this ordering change should be fine, but we do need to support both pre and post triggering in the future (probably by changing SND_SOC_DPCM_TRIGGER_POST etc to be a mask that can be ORed).

**Comment by libinyang** [comment]
Remove RFC from the title.

I have updated the patches, which now add the new feature that BE can require the trigger order. The FE trigger order requirement is higher than BE trigger order requirement. Please help review.

**Comment by RanderWang** [comment]
Now  bxt_pcm512x.c is changed, so all other machine drivers using hda should be refined ?

**Comment by libinyang** [comment]
> Now bxt_pcm512x.c is changed, so all other machine drivers using hda should be refined ?

Yes, after we are OK with this solution, I will add more machine drivers.

File: sound/soc/sof/intel/hda-dai.c [hunk 3]
```diff
@@ -165,17 +165,33 @@ static int hda_link_hw_params(struct snd_pcm_substream *substream,
...
 	switch (cmd) {
 	case SNDRV_PCM_TRIGGER_RESUME:
-	case SNDRV_PCM_TRIGGER_START:
 	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
+		ret = restore_hw_params(substream, dai);
+		if (ret) {
+			dev_err(dai->dev, "restore hw_params fails\n");
+			return ret;
+		}
+		/* no break, continue to start */
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-dai.c:193]: need to add a /* fall-through */ comment to remove warnings.

File: sound/soc/intel/boards/bxt_pcm512x.c [hunk 4]
```diff
@@ -204,6 +204,7 @@ static struct snd_soc_dai_link dailink[] = {
 		.init = broxton_hdmi_init,
 		.dpcm_playback = 1,
 		.no_pcm = 1,
+		.trigger[0] = SND_SOC_DPCM_TRIGGER_POST,
```
Comment (plbossart) [inline sound/soc/intel/boards/bxt_pcm512x.c:207]: @libinyang If we are changing the behavior for the iDisp link, then you need to test all configurations with the SST/cAVS driver. we *cannot* use a solution for iDisp that works only for SOF. So please make sure this is tested on multiple platforms before we jump to this different trigger mode.

Comment (plbossart) [reply sound/soc/intel/boards/bxt_pcm512x.c:207]: specifically all machines drivers I looked at use this for iDisp:
.trigger = {
			SND_SOC_DPCM_TRIGGER_POST, SND_SOC_DPCM_TRIGGER_POST},

so why do we only set the first element?

**Comment by ranj063** [comment]
@libinyang this seems like quite a round about way of setting the trigger order for the FE dai link.

I think what we should do instead in just set the trigger order in topology. I understand there's no support for setting the trigger order for PCM's topology today. But can we possibly look into extending the struct snd_soc_tplg_pcm {} to add the trigger order. Then, we can use it to set the trigger order in static int soc_tplg_fe_link_create() during topology parsing. What do you think?

File: sound/soc/sof/intel/hda-dai.c (hunk 3 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-dai.c:193]: @libinyang why do we need to restore hw_params for PAUSE_RELEASE?

**Comment by ranj063** [comment]
we could possible look into extending the SND_SOC_TPLG_LNK_FLGBIT_* which are the dai link flags to set the trigger order which would keep changes to minimum

File: sound/soc/sof/intel/hda-dai.c (hunk 3 above)
Comment (libinyang) [reply sound/soc/sof/intel/hda-dai.c:193]: copy from pcm.c sof_pcm_trigger(). I'm OK to remove it.

Comment (libinyang) [reply sound/soc/sof/intel/hda-dai.c:193]: OK. Thanks for it. "fall-through" is better than my comment.

> need to add a /* fall-through */ comment to remove warnings.

File: sound/soc/intel/boards/bxt_pcm512x.c (hunk 4 above)
Comment (libinyang) [reply sound/soc/intel/boards/bxt_pcm512x.c:207]: > specifically all machines drivers I looked at use this for iDisp:
> .trigger = {
> SND_SOC_DPCM_TRIGGER_POST, SND_SOC_DPCM_TRIGGER_POST},
> 
> so why do we only set the first element?

@plbossart 
I only set for the first element because there is no capture for HDMI audio.

Comment (libinyang) [reply sound/soc/intel/boards/bxt_pcm512x.c:207]: @plbossart Currently, I only change the setting of the platform bxt_pcm512x.c. The default mode is still the same as before. This means other platforms will not be impacted. And yes, I will test on other platforms to see if others need such setting.

> @libinyang If we are changing the behavior for the iDisp link, then you need to test all configurations with the SST/cAVS driver. we _cannot_ use a solution for iDisp that works only for SOF. So please make sure this is tested on multiple platforms before we jump to this different trigger mode.

**Comment by libinyang** [comment]
@ranj063 Actually, it is not a workaround. It is a feature. I believe both FE and BE should have the choice to change the priority. It depends on the real HW where we should set the priority.

If FE needs a special priority, we can set in FE. If BE needs a special priority, we can set in BE. For example,
if there are 2 BE connect to the FE (FE-BE connection allow this situation). FE doesn't care about the priority and BE1 doesn't care about the priority. But BE2 needs a special priority. We must allow BE2 to set the priority and change the behavior. 

In my design:
1. If FE needs a special priority, all BEs should follow this FE.
2. If FE doesn't care about the priority, BEs need negotiate for the priority. My current negotiation is very easy, take the first served BE priority (we can improve the negotiation later).
3. If both FE and BE don't care, we will use default mode, which is the same as before. And nothing will be impacted.

In this case, it's actually the HDA link DMA (belongs to BE) requirement that need BE need be triggered firstly. So it's the requirement from BE.

> @libinyang this seems like quite a round about way of setting the trigger order for the FE dai link.
> 
> I think what we should do instead in just set the trigger order in topology. I understand there's no support for setting the trigger order for PCM's topology today. But can we possibly look into extending the struct snd_soc_tplg_pcm {} to add the trigger order. Then, we can use it to set the trigger order in static int soc_tplg_fe_link_create() during topology parsing. What do you think?

**Comment by libinyang** [comment]
@ranj063 Thanks for the reference. I think this is another feature. And I will implement it after this bug is fixed.
> we could possible look into extending the SND_SOC_TPLG_LNK_FLGBIT_* which are the dai link flags to set the trigger order which would keep changes to minimum

**Comment by ranj063** [comment]
> if there are 2 BE connect to the FE (FE-BE connection allow this situation). FE doesn't care about the priority and BE1 doesn't care about the priority. But BE2 needs a special priority. We must allow BE2 to set the priority and change the behavior.

@libinyang what would be the intended behaviour in this case if the 2 BE dai link's set opposite trigger orders?

**Comment by libinyang** [comment]
@ranj063 Mostly, this should not happen. If it happens, it mean either HW or SW is designed wrongly. Anyway, we should handle this situation. In my design, we will use the first served BE's order.

> 
> @libinyang what would be the intended behaviour in this case if the 2 BE dai link's set opposite trigger orders?

**Comment by ranj063** [comment]
@libinyang I'm not against this idea but I feel that this is too disruptive. For our purposes, it might suffice to set the trigger order for the HDMI pipeline PCM's which will then be assigned to the FE DAI links that will be created when the PCM elems are parsed. This way we will also have the flexibility in the future to support pre/post/bespoke triggers as well.

In general, I currently see no reference to the be dai link trigger being used in dpcm and I am worried about the feedback from community. 

@lgirdwood what do you think?

**Comment by libinyang** [comment]
@ranj063 Actually, we still have the chance of setting order in FE, which I'm planning to support this feature. But if FE doesn't care, we can let BE to decide. 

> @libinyang I'm not against this idea but I feel that this is too disruptive. For our purposes, it might suffice to set the trigger order for the HDMI pipeline PCM's which will then be assigned to the FE DAI links that will be created when the PCM elems are parsed. This way we will also have the flexibility in the future to support pre/post/bespoke triggers as well.
> 
> In general, I currently see no reference to the be dai link trigger being used in dpcm and I am worried about the feedback from community.
> 
> @lgirdwood what do you think?

**Comment by ranj063** [comment]
> @ranj063 Actually, we still have the chance of setting order in FE, which I'm planning to support this feature. But if FE doesn't care, we can let BE to decide.

How about we use this for the fix and you float your idea to propagate trigger order from BE dai link to FE DAI link upstream?

**Comment by libinyang** [comment]
@ranj063 I do think that the trigger order requirement is from BE. So we should set it in BE.

@plbossart @lgirdwood @RanderWang What do you think?

> How about we use this for the fix and you float your idea to propagate trigger order from BE dai link to FE DAI link upstream?

**Comment by plbossart** [comment]
@libinyang I am not clear on why we need specific changes that weren't needed for the cAVS driver - despite using the same DPCM framework. This will require more reviews and I will have to push until next week to look into this. Please add clarifications in the mean time.

**Comment by libinyang** [comment]
@plbossart As you can see, cAVS uses SND_SOC_DPCM_TRIGGER_POST in many machine drivers in sound/soc/intel/boards/*. It defines the order in the FE link. However, the cAVS is not precise. As you know, this trigger order requirement is from BE link. So we should set it in BE link. This is what my patch is doing.

> @libinyang I am not clear on why we need specific changes that weren't needed for the cAVS driver - despite using the same DPCM framework. This will require more reviews and I will have to push until next week to look into this. Please add clarifications in the mean time.

**Comment by plbossart** [comment]
> @plbossart As you can see, cAVS uses SND_SOC_DPCM_TRIGGER_POST in many machine drivers in sound/soc/intel/boards/*. It defines the order in the FE link. However, the cAVS is not precise. As you know, this trigger order requirement is from BE link. So we should set it in BE link. This is what my patch is doing.

Sorry, lost in translation here. What does 'not precise' mean here? either the sequence is correct or it's not, and the order is correct or it's not. And the trigger is also set in the BE.

**Comment by ranj063** [comment]
> Sorry, lost in translation here. What does 'not precise' mean here? either the sequence is correct or it's not, and the order is correct or it's not. And the trigger is also set in the BE.

@plbossart as far as I could tell, the trigger in the Intel machine drivers only sets the trigger order for the FE dai links. This is where my concern was from. It makes sense to add it to the FE dai links because this is what is used in dpcm and I think it makes the most sense to me at least.

**Comment by plbossart** [comment]
> > Sorry, lost in translation here. What does 'not precise' mean here? either the sequence is correct or it's not, and the order is correct or it's not. And the trigger is also set in the BE.
> 
> @plbossart as far as I could tell, the trigger in the Intel machine drivers only sets the trigger order for the FE dai links. This is where my concern was from. It makes sense to add it to the FE dai links because this is what is used in dpcm and I think it makes the most sense to me at least.

Exactly my point. If the machine drivers set this for FE so far, then WTH do we need this for SOF in the backends? what is so specific about SOF?

**Comment by ranj063** [comment]
> Exactly my point. If the machine drivers set this for FE so far, then WTH do we need this for SOF in the backends? what is so specific about SOF?

@plbossart Nothing SOF specific at all. It is most intuitive to add a trigger to the FE dai link even in SOF. @libinyang what say?

**Comment by libinyang** [comment]
What I think is that cAVS is not precise. It should also set it in BE, although it works now. Setting it in FE is not a perfect option. The trigger order requirement is from BE, so we should set the trigger order flag in BE. Anyway, if both of you are agree to use FE, I'm OK to switch to FE. I really do think setting in BE is a better solution.

> 
> @plbossart Nothing SOF specific at all. It is most intuitive to add a trigger to the FE dai link even in SOF. @libinyang what say?

**Comment by ranj063** [comment]
> What I think is that cAVS is not precise. It should also set it in BE, although it works now. Setting it in FE is not a perfect option. The trigger order requirement is from BE, so we should set the trigger order flag in BE. Anyway, if both of you are agree to use FE, I'm OK to switch to FE. I really do think setting in BE is a better solution.

@libinyang The trigger is always initiated at the FE, so in my opinion the trigger order should be specified for the FE dai link. Setting it for the BE dai link doesnt make it quite intuitive. I understand though where you are coming from with your argument ie the link hw_params need to be set before the host hw_params.

**Comment by libinyang** [comment]
@ranj063 Actually, as you see the link hw_params is for the BE, and the requirement of trigger order is from hda link, which belongs to BE. This is why I want to set it in BE. If the requirement is from host DMA, which belongs to FE, I would like to set it in FE.

> 
> @libinyang The trigger is always initiated at the FE, so in my opinion the trigger order should be specified for the FE dai link. Setting it for the BE dai link doesnt make it quite intuitive. I understand though where you are coming from with your argument ie the link hw_params need to be set before the host hw_params.

**Comment by libinyang** [comment]
I start to implement the FE trigger order now. Before I start, let's sync up on one question. As Ranjani mentioned in her previous comment, there is "SND_SOC_TPLG_LNK_FLGBIT_*" as a reference. It is stored in the pcm->flags. To implement trigger order, do we store it in pcm->flags or we use new members such as trigger_order[2]? I prefer to introduce new struct snd_soc_tplg_pcm member named trigger_order[2]. What do you think? This means we will change the struct snd_soc_tplg_pcm .

 @ranj063 @plbossart

**Comment by ranj063** [comment]
> I start to implement the FE trigger order now. Before I start, let's sync up on one question. As Ranjani mentioned in her previous comment, there is "SND_SOC_TPLG_LNK_FLGBIT_*" as a reference. It is stored in the pcm->flags. To implement trigger order, do we store it in pcm->flags or we use new members such as trigger_order[2]? I prefer to introduce new struct snd_soc_tplg_pcm member named trigger_order[2]. What do you think? This means we will change the struct snd_soc_tplg_pcm .
> 
> @ranj063 @plbossart

@libinyang adding the trigger order members to struct snd_soc_tplg_pcm doesnt seem right as this isnt really a pcm attribute. The TPLG_LNK_FLKBIT* looks like the right way to do it. @plbossart do you have an opinion on this?

**Comment by libinyang** [comment]
@ranj063 It's my first time to touch this area. Please help pointer it out if I'm wrong.

Look into the tplg parse code in soc-topology.c. The SND_SOC_TPLG_LNK_FLGBIT_xxx is set in function set_link_flags(). set_link_flags() is called from 2 functions: 
1. soc_tplg_fe_link_create(): this is used to handle the FE link
2.soc_tplg_link_config(): this is used to handle the BE link

I think in our case, it's FE link. So we should care about soc_tplg_fe_link_create(). In soc_tplg_fe_link_create(), it calls set_link_flags(link, pcm->flag_mask, pcm->flags). So it uses pcm to save the flag. "pcm" is struct snd_soc_tplg_pcm *pcm. struct snd_soc_tplg_pcm *pcm comes from tplg.

So in our case, we should:
1. add the trigger flag in m4 file.
2. change the alsa-lib to support trigger flag. and build the tplg to support trigger flag.
3. When parsing tplg binary, kernel driver will use the updated "struct snd_soc_tplg_pcm" to parse the new tplg.
4. parse the trigger flag in "struct snd_soc_tplg_pcm" and save it into the  snd_soc_dai_link.

**Comment by ranj063** [comment]
> @ranj063 It's my first time to touch this area. Please help pointer it out if I'm wrong.
> 
> Look into the tplg parse code in soc-topology.c. The SND_SOC_TPLG_LNK_FLGBIT_xxx is set in function set_link_flags(). set_link_flags() is called from 2 functions:
> 
> 1. soc_tplg_fe_link_create(): this is used to handle the FE link
>    2.soc_tplg_link_config(): this is used to handle the BE link
> 
> I think in our case, it's FE link. So we should care about soc_tplg_fe_link_create(). In 
@libinyang Yes, this is correct.

soc_tplg_fe_link_create(), it calls set_link_flags(link, pcm->flag_mask, pcm->flags). So it uses pcm to save the flag. "pcm" is struct snd_soc_tplg_pcm *pcm. struct snd_soc_tplg_pcm *pcm comes from tplg.
> 
> So in our case, we should:
> 
> 1. add the trigger flag in m4 file.
> 2. change the alsa-lib to support trigger flag. and build the tplg to support trigger flag.
> 3. When parsing tplg binary, kernel driver will use the updated "struct snd_soc_tplg_pcm" to parse the new tplg.
> 4. parse the trigger flag in "struct snd_soc_tplg_pcm" and save it into the  snd_soc_dai_link.

perfect. this is exactly what we need. Thanks!

**Comment by libinyang** [comment]
@ranj063 Thanks for confirmation.

**Comment by libinyang** [comment]
@ranj063 @plbossart @mengdonglin 
I have worked out the patch for the kernel and the patch for alsa-lib. After testing, the patches work. The m4 patch is not ready.
However, as Mengdong mentioned in the email, we should notice the ABI changing. This may be challenged by community. I have got an idea for the ABI compatibility, I will discuss with community for the idea to see if they are OK with this idea.

**Comment by ranj063** [comment]
@libinyang @plbossart here's comment from Takashi about our PM flow and I think it makes sense.
Without the INFO_RESUME set in hw_params, he says that we should setup the hw_params in prepare(). I will submit the change for this. Maybe, we wont need to switch the trigger order with this?

"So, SOF mandates the re-setup of the hw params in BE at resume?
Then the right place to do is in the prepare callback.  Without
SNDRV_PCM_INFO_RESUME flag, the PCM core won't trigger RESUME, but
returns -ENOSYS.  Then user-space will do prepare, and start with the
normal trigger.

You'd need a flag indicating the BE hw_params resetup, and at the
beginning of prepare function, the flag is checked and the (internal)
hw_params is done accordingly."

**Comment by plbossart** [comment]
@libinyang I am not going to touch all these HDMI changes for the v6 upstream contributions. I don't have the time to look into the details, I don't understand if this is an SOF-specific issue or it doesn't also work with cAVS-based solutions, and last because they depend on changes that need to be approved by higher powers.

**Comment by ranj063** [comment]
> @libinyang I am not going to touch all these HDMI changes for the v6 upstream contributions. I don't have the time to look into the details, I don't understand if this is an SOF-specific issue or it doesn't also work with cAVS-based solutions, and last because they depend on changes that need to be approved by higher powers.

@plbossart i dont we need this PR (not all of it) anymore. I have implemented some changes to our pcm resume flow based on Takashi's feedback. With that, the only 2 patches we need from @libinyang are 
1.  save fixed up hw_params for be substream
2. set up device link between machine drv and hdmi dev

I will submit my PR soon.

**Comment by plbossart** [comment]
@libinyang it's my understanding that this PR can be closed?
